sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -339,8 +339,24 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
       case SqlBaseParser.SELECT =>
         // Regular select
 
+        // Broadcast hints
+        var withBroadcastedTable = relation
```

Comment:
A few thoughts here:

- Should we introduce a separate LogicalPlan (like `With`) and move this logic into an Analyzer rule? The reason that I am proposing this is because we are transforming the `LogicalPlan` here. It is simpler to do it here though.
- Lets move the logic here into a separate method. Name it `withHints`, model it after the other `with*` functions, and use `optionalMap(hint)(withHint)`.